### ISSUE TITLE
Role of XML namespaces in MNX

### ISSUE THREAD

**Issue by joeberkovitz** [w3c/mnx#77]
MNX has at least 3 separate areas defined so far:

- Container (Integration)
- Western (CWMNX)
 -Generic (GMNX)

Additional areas are expected to be added for future semantic dialects, of which Western/CWMNX happens to be the first.

What role do namespaces play in qualifying and scoping these various areas of MNX? This hasn't been discussed much to date except in #74.  They could play a role in:
- splitting the schema into separate, logically consistent portions
- leveraging XML's own extension mechanism to make it easy to add new semantic dialects
- mitigate the collisions between dialects using frequently occurring words as element names (e.g `<sequence>`)

Note that there is definitely some interpenetration of elements foreseen already in the draft spec.  For example, the Western/CWMNX schema can incorporate low-level performance data taken from the Generic/GMNX schema.

**Comment by joeberkovitz** [comment]
The naming of the top-level elements of the various dialects (e.g. `<gmnx>` or `<mnx-western>`) also belongs to this thread since it interacts strongly with the approach to namespacing, if any.

**Comment by shoogle** [comment]
Would it be possible to have a parent namespace with abstract music definitions that are dialect-agnostic, that are then made more concrete in the sub-namespaces?

For example, there is no audible difference between C-sharp and D-flat (at least in equal temperament), so enharmonic spelling could be a property of the dialect (CWMN) rather than the abstract base namespace, which would be reserved for properties that have physical significance.

```xml
<sequence xmlns="http://example.com/abstract" xmlns:cwmn="http://example.com/cwmn">
  <event duration="2beats">
    <note octave="4" pitch="1" cwmn:spelling="flat">
  </event>
</sequence>
```

Enharmonic spelling can sometimes be physically significant, but not always. It could be considered __semi-semantic__, as it is not purely semantic or purely stylistic in nature.

Other examples of semi-semantic elements in CWMN:

- Tied notes
  - There is no audible difference between a minim compared to two tied crotchets
  - Yet ties are often used to communicate stress (like beaming) and exist in other dialects
- Key signatures, accidentals and enharmonics
  - The key of C Major appears to be special when in reality all keys are equal and only the melody is significant
- The "denominator" of a time signature
  - e.g. No audible difference between a piece written in 2/4 vs 2/2
- Compound time signatures
  -  e.g. No audible difference between a piece in 6/8 vs 2/4 with triplets.

One advantage of separating the "semi-semantic" properties into other namespaces is that they can be easily ignored during conversion between dialects, but if two dialects share the concept of a tied note (for example) then they can use the data from the other namespace as a hint during conversion.

The use of namespaces would also enable data for multiple dialects to be stored in a single file.

**Comment by evanlenz** [comment]
TL;DR: Use one namespace URI for all MNX elements

In case XML namespaces are still being considered, I would advocate for a basic use of namespaces, not to disambiguate between closely-related specs that you control, but rather between MNX and other vocabularies. In other words, decide on just one namespace URI for MNX and then use it for all MNX element names in all MNX specs/profiles.

Having worked with XML databases like MarkLogic (in which I look forward to querying MNX data someday!), I know that having a namespace URI (especially with generic-sounding element names like "sequence", "part", "global", etc.), will be useful in avoiding name collisions, since by default it indexes elements just by their QName.

That said, I would recommend against a more ambitious use of namespaces, e.g. a different namespace URI for each profile or even version. I wrote an introduction to XML namespaces a while back and included a [short discussion on their motivation](http://lenzconsulting.com/namespaces/#motivation), in case anyone wants to dig a little deeper into where I'm coming from on this.

P.S. Nice to meet everyone :-)

**Comment by notator** [comment]
@evanlenz: Thanks for joining us. Happy New Year!

Namespaces are going to be very relevant once we start developing a formal _Schema_ for MNX. That's something we could start doing fairly soon, and it might even be an elegant answer to @adrianholovaty's search for a better way to develop the MNX Specification.

Our current reference document, [The Draft Spec](https://w3c.github.io/mnx/specification/common/), has been a great place to start (it contains some golden nuggets that came from @joeberkovitz' experience with MusicXML), but it is now unsatisfactory for three main reasons:
1. It is difficult to edit.
2. It is impossible to distinguish between the parts that have been agreed, and those that are still controversial.
3. It can't be used to formally verify MNX documents.

The best way to find out what we've currently agreed is to look at the concrete code examples in [_MNX by Example_](https://w3c.github.io/mnx/by-example/). The _Schema_ would build on those, but also be informed by the more generalised definitions of the relevant elements in the Draft Spec. Defining a succinct, formal _Schema_ would lead to progress because any deficiencies in the (less formal) Draft Spec would be discovered and eliminated in the process.

In other words: The _Schema_ and _MNX by Example_ could be kept consistent, while we gradually retire the Draft Spec.

@evanlenz: _MNX by Example_ only contains examples of "basic CWMN", but this CG has also decided to create other, related formats. As I understand it, you are recommending _against_ using namespaces to differentiate between different formats. But are you saying that its okay to distinguish element types using sub-namespaces (e.g. between `<cmn1900:sequence>` and `<cmn2000:sequence>`)? It would be great if you could help us get the _Schema_ started out on the right foot! :-)

Whatever happens, the existence of a developing _Schema_ would be the prerequisite for implementing more advanced ideas such as those described by @shoogle [above](https://github.com/w3c/mnx/issues/77#issuecomment-447501176), so I think this is the way to go.

See also: Issue #73

**Comment by shoogle** [comment]
My [idea above](https://github.com/w3c/mnx/issues/77#issuecomment-447501176) seemed good at the time, but now I think it would make things way too complicated for implementors. I now agree with @evanlenz that everything in the Common spec should be in one namespace.

While would be nice to separate different parts of the spec into different namespaces, the fact is that namespaces are poorly supported in major XML parsers (e.g. [Python ElementTree](https://stackoverflow.com/a/14853417)) and even in the original specification their behaviour is counter-intuitive (especially for attributes, which [don't inherit the namespace of the "parent" element](http://xmlplease.com/xml/attributexmlns/#s2.)). If we put everything in the default namespace then developers don't have to deal with these complexities.

If we want to show that certain elements and attribute are related then a better solution would be to use a prefix like  `cwmn_` as a kind of "visual namespace" rather than a real XML namespace like `cwnm:`.

```xml
<!-- Namespace -->
<note octave="4" pitch="1" cwmn:spelling="flat">

<!-- Prefix -->
<note octave="4" pitch="1" cwmn_spelling="flat">
```

The prefix would be enough to show that certain elements and attributes are related without developers having to understand all the ins and outs of actual XML namespaces.

**Comment by notator** [comment]
I've used namespaces, and created a few simple schemas in the past, but don't _really_ understand what @evanlenz and @shoogle are saying. A concrete example would help.
If a Schema were to be derived from the [Hello World](https://w3c.github.io/mnx/by-example/#helloworld) example, how would it look?

**Comment by shoogle** [comment]
@notator, TLDR: Don't use XML namespaces (except the default namespace).

**Comment by notator** [comment]
@shoogle: Thanks for the clarification. So this issue is settled, and should be closed. We should continue in #73...

**Comment by adrianholovaty** [comment]
Marking this as closed, for the sake of cleaning up our issue database. This is very outdated — it deals with the older XML version of MNX.
